drivers/hmis/app/models/hmis/form/instance_project_match.rb

```diff
@@ -0,0 +1,65 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# determine if project is a valid match for a form instance. The match is ranked so that when there are multiple form
+# definitions that match a project, we can take the definition with the best rank
+#
+class Hmis::Form::InstanceProjectMatch
+  include Memery
+  attr_accessor :project, :instance
+
+  # match types ordered by rank. Lower rank is better
+  RANKED_MATCHES = [
+    PROJECT_MATCH = 'project'.freeze,
+    ORGANIZATION_MATCH = 'organization'.freeze,
+    PROJECT_TYPE_AND_FUNDER_MATCH = 'project_type_and_funder'.freeze,
+    PROJECT_TYPE_MATCH = 'project_type'.freeze,
+    PROJECT_FUNDER_MATCH = 'project_funder'.freeze,
+    DEFAULT_MATCH = 'default'.freeze,
+  ].freeze
+  MATCH_RANKS = RANKED_MATCHES.each_with_index.to_h.freeze
+
+  def initialize(instance:, project:)
+    self.instance = instance
+    self.project = project
+  end
+
+  def rank
+    MATCH_RANKS[match]
+  end
+
+  def valid?
+    !!match
+  end
+
+  protected
+
+  # match to project. Order is significant, should return the best ranked match
+  memoize def match
+    if instance.entity_type
+      case instance.entity_type
+      when Hmis::Hud::Project.sti_name
+        return PROJECT_MATCH if instance.entity_id == project.id
+      when Hmis::Hud::Organization.sti_name
+        return ORGANIZATION_MATCH if instance.entity_id == project.organization.id
+      else
+        # entity type is specified but doesn't match
+        return
+      end
+    end
```

Comment:
won't this "fall through"  down to line 54 if there is an entity type specified but the current one doesn't match? if there is an entity specified and it doesn't match, we want to return immediately.

```suggestion
    if instance.entity_type
      case instance.entity_type
      when Hmis::Hud::Project.sti_name
        return PROJECT_MATCH if instance.entity_id == project.id
      when Hmis::Hud::Organization.sti_name
        return ORGANIZATION_MATCH if instance.entity_id == project.organization.id
      end
      
      return  # entity type is specified but doesn't match
    end
```

Reply:
Good catch! I tightened up the conditions here